src/modules/change-email/controller.js

```diff
@@ -0,0 +1,130 @@
+ * @param  {Object}  request
+ * @param  {Object}  h
+ */
+const postGenerateSecurityCode = async (request, h) => {
+  const { verificationId, newEmail } = request.payload;
+  try {
+    const { data: { verificationCode } } = await addNewEmailAddressIDM(verificationId, newEmail);
+
+    // Send email to new email address with security code
+    const notifyTemplate = findByMessageRef('email_change_verification_code_email');
+    await notify.send(notifyTemplate, { verification_code: verificationCode }, newEmail);
+  } catch (error) {
+    if (error.message === 'Email address already in use') {
+      sendEmailAddressInUseNotification(newEmail);
+    }
+    return error;
+  }
+};
+
+/**
+ * check security code, log event, send confirmation email to old email address
+ * @param  {Object}  request
+ * @param  {Object}  h
+ */
+const postChangeEmailAddress = async (request, h) => {
+  const { securityCode, entityId } = request.payload;
+  const { userId, userName } = request.defra;
+
+  try {
+    const { data: { newEmail } } = await verifySecurityCode(userId, securityCode);
+
+    // Updates CRM individual entity with new email address
+    await updateEmailInCRM(entityId, newEmail);
+
+    // Create event object
+    const evt = createEventObject(userName, entityId, newEmail, userId);
+
+    // Logs event in events table with old/new email address and user ID
+    await event.repo.create(evt);
+  } catch (error) {
+    logger.error('Email change error', error);
+    if (error.name === 'EmailChangeError') return error;
+  }
+};
+
+/**
+ * call IDM to check user credentials and create email change record
+ */
+const createEmailChangeRecordIDM = (userId, password) => {
+  const url = `${process.env.IDM_URI}/user/change-email-address/start`;
+  return helpers.serviceRequest.post(url, { userId, password });
+};
+
+/**
+ * call IDM to update record with new email address
+ */
+const addNewEmailAddressIDM = (verificationId, newEmail) => {
+  const url = `${process.env.IDM_URI}/user/change-email-address/create-code`;
+  return helpers.serviceRequest.patch(url, { verificationId, newEmail });
+};
+
+/**
+ * check if record matching userId and securityCode exists
+ * @param  {Int} userId
+ * @param  {Int} securityCode
+ */
+const verifySecurityCode = (userId, securityCode) => {
+  const url = `${process.env.IDM_URI}/user/change-email-address/complete`;
+  return helpers.serviceRequest.post(url, { userId, securityCode });
+};
+
+/**
+ * create event object to be inserted into event log
+ */
+const createEventObject = (userName, entityId, newEmail, userId) => {
+  return event.create({
+    event_id: uuid(),
+    type: 'user-account',
+    subtype: 'email-change',
+    issuer: userName,
+    entities: [entityId],
+    metadata: {
+      'old-email': userName,
+      'new-email': newEmail,
+      'user-id': userId
+    }
+  });
+};
+
+/**
+ * update CRM db row for entity with new email for entity_nm
+ */
+const updateEmailInCRM = async (entityId, newEmail) => {
```

Comment:
As above for IDM connector code, this should probably live in lib/connectors/crm